### PR TITLE
projectm 4.0.0

### PR DESCRIPTION
**Pull request by chenrui333**
Created with `brew bump-formula-pr`.<details>
  <summary>release notes</summary>
  <pre>This is the final 4.0 release of libprojectM.

It's been about two years after the last 3.1 release, but the projectM core team and community worked hard to make projectM better for both end users and developers integrating the library into their own applications.

Since this release is mainly aimed at developers, we will only attach source packages to this release. If you want pre-built libraries, please download the latest artifacts from the [automated GitHub Actions builds](https://github.com/projectM-visualizer/projectm/actions?query=branch%3Amaster).

As always, if you find bugs or have feature suggestions or requests, please feel free to [create an issue](https://github.com/projectM-visualizer/projectm/issues/new/choose) or post [in the discussions board](https://github.com/orgs/projectM-visualizer/discussions/712). We're also available for chat on our Discrod server, which is linked on the project page.

### Target Audience: Developers
This release is mainly aimed at developers integrating projectM into their applications. The main focus of this major release was to provide a new, stable API based on pure C functions, which will make it easier to create bindings for other languages as well as using it as a shared library. We also fixed many crashes and improved the overall visual quality and compatibility of presets to some degree.

The API will remain stable and forward-compatible until the next major release. Any changes made will only include new functions, so the ABI will always stay forward-compatible. This means an application built and linked against a specific 4.x version will build and run with any libprojectM with a major version of 4 and the same or higher minor version. If an application uses libprojectM as a shared library, no relinking is required at all - simply swap the library file with the new one and it should work.

### What About Users?
If you are an end user looking for the standalone visualizer or the Apple Music plug-in, this release isn't very useful as it only contains the projectM development libraries and headers, but no frontend or executable you can run.

Since modernizing the projectM ecosystem is a big undertaking, we've decided to postpone the major visual and UI improvements to a later point, first releasing the core library so developers can adopt the new programming interface.

You can expect a new frontend application and great Milkdrop compatibility improvements on the upcoming minor 4.1 release, which we will make available together with the updated frontends. Until then, you can use these development releases, which contain most of the changes in this 4.0 release:

- Standalone application for Windows/macOS/Linux: https://github.com/kblaschke/frontend-sdl2/releases
- Apple Music plug-in for macOS: https://github.com/kblaschke/frontend-music-plug-in/releases/tag/v3.0-pre1

## Changes Since 3.1.12

To better distiguish what's part of this release and what's not, we have compiled a changelog and a list with issues that will not be fixed or implemented in the initial 4.0 version:

### User-Related Changes
- Improved waveform and shape line rendering to match Milkdrop's way of drawing.
- Fixed some bugs rendering the different waveform types.
- Added support for instanced shapes. In previous releases, only a single shape was rendered for each shapecode block, breaking many presets.
- Fixed the built-in noise textures used by preset shaders by reimplementing Milkdrop's generator code. Noise textures now have four different color channels instead of having the same value on RGB and 1.0 only in the alpha channel.
- Fixed texture coordinates in textured shapes.
- Fixes in the preset comp/warp shader parsing code, leading to fewer broken presets.
- Fixed numerous crashes when parsing invalid preset expressions or shaders.
- Many more small stability and quality improvements.

### Developer-Related Changes
- Replaced the platform-specific build files with CMake.
- Removed all non-library-related applications and files from the repository. This includes all presets except the development test presets, any user-faced frontends and other old code. The previous SDL frontend was retained as a developer-only test UI, but is not built by default and also not installed alongside the library.
- Added a new C API, hiding all C++ types by default. This will make it easier to integrate projectM into other languages.
- Added proper export defines to all functions. libprojectM can now be used as a shared library on all supported platforms, including Windows.
- Moved preset playlist support to a separate, optional library. The core library now only supports loading a single preset plus a second one during transitions. The new playlist library supports loading multiple paths/files and can apply .gitignore-style filter patterns. Rating support has been removed though.
- Removed all text overlays.
- Replaced the key handler with specific API functions.
- Replaced the settings struct with getter and setter API functions.
- Removed built-in configuration file support. All non-default settings now need to be passed in from the embedding application.
- Improved and modernized the audio data interface.
- Enabled OpenMP support on all platforms.
- Support multiple texture search paths.
- Added support for loading presets from a character buffer instead of a file.
- Added callbacks for preset switches and loading failures in the core library.
- Fixed a crash that occurred if OpenGL wasn't initialized properly. `projectm_create()` will now simply return NULL in this case.

## Issues Remaining Unresolved in the 4.0.0 Release

These issues will be fixed or implemented after the 4.0 release, as they don't affect the API backward compatibility:

- Supporting missing expression features like `megabuf` and `loop`.
- Fixing the issue projectM is loading all textures into (V)RAM on startup instead of only loading those which are used by the current preset.
- Improve rendering quality on high-res displays by gradually scaling line thickness.
- Adding more debugging functions to analyze rendering internals and preset expression values on the application side.
</pre>
</details>

### CODE REVIEW

**Comment by chenrui333** [comment]
`Failed to execute: ./configure`

**Comment by hartwork** [comment]
@chenrui333 4.x.x is using CMake now. Also I'd expect this bump to break `qmmp.rb` unless 3.x.x and 4.x.x are made to live side by side or QMMP is made to work with 4.x.x. I haven't checked yet.

**Comment by hartwork** [comment]
@chenrui333 have you checked whether it breaks qmmp? I believe this needs a versioned package with 3.x and 4.x side by side rather than a replacement bump. Thanks for your consideration :pray:

**Comment by chenrui333** [comment]
let me check on it today

**Comment by hartwork** [comment]
> let me check on it today

@chenrui333 I had a quick look myself now, QMMP is not going to work with projectM 4 unpatched, e.g. because the pkg-config name now is `projectM-4` rather than the `libprojectM` of <4 that QMMP is looking for:

```console
# fgrep projectM qmmp-2.1.3/src/plugins/Visual/projectm/CMakeLists.txt
# projectM
  pkg_search_module(PROJECTM libprojectM>=2.0.0)
#  MESSAGE(WARNING "Unsuppoted OpenGL implementation, projectM disabled")
pkg_search_module(PROJECTM_31 libprojectM>=3.1.0)
          PATHS "/usr/share/projectM/" "/usr/local/share/projectM/")
MESSAGE(STATUS "Found projectM configuration: ${PROJECTM_CONFIG}")
```

So I believe it takes more than a rebump and a rebuilt. What do you think?

**Comment by chenrui333** [comment]
> > let me check on it today
> 
> @chenrui333 I had a quick look myself now, QMMP is not going to work with projectM 4 unpatched, e.g. because the pkg-config name now is `projectM-4` rather than the `libprojectM` of <4 that QMMP is looking for:
> 
> ```
> # fgrep projectM qmmp-2.1.3/src/plugins/Visual/projectm/CMakeLists.txt
> # projectM
>   pkg_search_module(PROJECTM libprojectM>=2.0.0)
> #  MESSAGE(WARNING "Unsuppoted OpenGL implementation, projectM disabled")
> pkg_search_module(PROJECTM_31 libprojectM>=3.1.0)
>           PATHS "/usr/share/projectM/" "/usr/local/share/projectM/")
> MESSAGE(STATUS "Found projectM configuration: ${PROJECTM_CONFIG}")
> ```
> 
> So I believe it takes more than a rebump and a rebuilt. What do you think?

looks like the build passed, let me see if qmmp side has issue tracking the projectm upgrade

**Comment by chenrui333** [comment]
indeed, there is an issue on this, https://sourceforge.net/p/qmmp-dev/tickets/1137/

**Comment by hartwork** [comment]
@chenrui333 I believe the build passes because it complete against the old projectM 3 not 4. Would that be possible?

**Comment by chenrui333** [comment]
I dont think so.

**Comment by hartwork** [comment]
> I dont think so.

@chenrui333 then the only explanation left is that the QMMP build system is okay with not finding projectM <4 and just doesn't build the projectM plugin at all.

**Comment by chenrui333** [comment]
```
  ==> pkg-config sdl2 --cflags --libs
  ==> /usr/bin/g++-11 -std=c++11 test.cpp -o test -I/home/linuxbrew/.linuxbrew/Cellar/projectm/4.0.0/include -L/home/linuxbrew/.linuxbrew/Cellar/projectm/4.0.0/lib -lprojectM-4 -lGL -D_REENTRANT -I/home/linuxbrew/.linuxbrew/include/SDL2 -L/home/linuxbrew/.linuxbrew/lib -Wl,-rpath,/home/linuxbrew/.linuxbrew/lib -Wl,--enable-new-dtags -lSDL2
  /usr/bin/ld: /home/linuxbrew/.linuxbrew/Cellar/projectm/4.0.0/lib/libprojectM-4.so: undefined reference to `glGetInternalformativ'
  collect2: error: ld returned 1 exit status
```

**Comment by chenrui333** [comment]
Looks like upstream has made a build patch, gonna try it out.
https://sourceforge.net/p/qmmp-dev/code/11031/

**Comment by hartwork** [comment]
@chenrui333 libvisual-projectm does not support projectM 4 as of today. I could try tackle that upstream but it would take a few days. A versioned formula would be the alternative I suppose. What do you think?

**Comment by chenrui333** [comment]
> @chenrui333 libvisual-projectm does not support projectM 4 as of today. I could try tackle that upstream but it would take a few days. A versioned formula would be the alternative I suppose. What do you think?

but does projectm would have any new releases for v3? (if not, we might just prefer waiting for your fix then)

**Comment by hartwork** [comment]
@chenrui333 I'm not sure I understand your idea. Could you rephrase or elaborate to help me understand?

**Comment by chenrui333** [comment]
> @chenrui333 I'm not sure I understand your idea. Could you rephrase or elaborate to help me understand?

yeah, we would make a versioned formula so that the upstream would support it

**Comment by chenrui333** [comment]
@hartwork I did not follow this PR closely, feel free to create a new one and carry it on.
